### ISSUE TITLE
🚀 Improve schemes generation

### ISSUE THREAD

**Issue by danyf90** [tuist/tuist#3975]
As discussed [here](https://github.com/tuist/tuist/pull/3967#discussion_r780699421) and on slack, the current scheme autogeneration logic generates more schemes than needed, this proposal aims at solving that issue by grouping targets in the same scheme.

## Problem

Often many targets are related and should be part of the same scheme as they belong to the same module.
For example, a module might be composed of (some of all):
- interface target
- implementation target
- testing helpers
- unit test target
- UITest target
- integration tests target
- demo target

Ideally, they should be grouped in the same scheme that has:
- build action -> interface, implementation, testing helpers
- test action -> unit tests, ui tests, integration tests
- run action -> demo target

### Target names assumptions

If we want to fix this in the automatic scheme generation logic, we could make the autogenerate logic more powerful and configurable to understand the type of target and grouping by their name. Int the example above the targets could have been named:
- interface target -> Feature
- implementation target -> FeatureImplementation
- testing helpers -> FeatureTesting
- unit test target -> FeatureTests
- UITest target -> FeatureUITests
- integration tests target -> FeatureIntegrationTests
- demo target -> FeatureDemo or FeatureApp

### Plugin

Instead of making the autogeneration logic more complex, we could define a plugin that allows to define feature modules at an higher level, similar to the helper we are using in the Tuist project description helper itself.

For example:
```
let feature = Feature(name: "Feature", hasTests: true, …)
…
let project = Project(
  …,
  targets: otherTargets + feature.targets,
  scheme: [
    …,
    feature.scheme
  ]
)
```

**Comment by ezraberch** [comment]
I think the problem is that the standard way to specify targets does not reflect the true relationship between them. Using the example from the tutorial:

```swift
let project = Project(
    name: "MyApp",
    organizationName: "MyOrg",
    targets: [
        Target(
            name: "MyApp",
            platform: .iOS,
            product: .app,
            bundleId: "io.tuist.MyApp",
            infoPlist: "Info.plist",
            sources: ["Sources/**"],
            resources: ["Resources/**"],
            headers: Headers(
                public: ["Sources/public/A/**", "Sources/public/B/**"],
                private: "Sources/private/**",
                project: ["Sources/project/A/**", "Sources/project/B/**"]
            ),
            dependencies: [
                /* Target dependencies can be defined here */
                /* .framework(path: "framework") */
            ]
        ),
        Target(
            name: "MyAppTests",
            platform: .iOS,
            product: .unitTests,
            bundleId: "io.tuist.MyAppTests",
            infoPlist: "Info.plist",
            sources: ["Tests/**"],
            dependencies: [
                .target(name: "MyApp")
            ]
        )
    ]
)
```

How about instead defining things this way:

```swift
let project = Project(
    name: "MyApp",
    organizationName: "MyOrg",
    targets: [
        Target(
            name: "MyApp",
            platform: .iOS,
            product: .app,
            bundleId: "io.tuist.MyApp",
            infoPlist: "Info.plist",
            sources: ["Sources/**"],
            resources: ["Resources/**"],
            headers: Headers(
                public: ["Sources/public/A/**", "Sources/public/B/**"],
                private: "Sources/private/**",
                project: ["Sources/project/A/**", "Sources/project/B/**"]
            ),
            dependencies: [
                /* Target dependencies can be defined here */
                /* .framework(path: "framework") */
            ]
            children: [
                Target(
                    product: .unitTests,
                    bundleId: "io.tuist.MyAppTests",
                    sources: ["Tests/**"],
                )
            ]
        )
    ]
)
```

This makes it clear what the relationship between the two is, allowing for them to be grouped in the same scheme. It also reduces duplication, as omitted fields can be inferred from the parent.

**Comment by danyf90** [comment]
@ezraberch I agree with you, but I think there are many way of organizing targets/modules and group them together, so it would be hard to make Tuist decide the "correct" way as it might differ among teams.

For that reason I think keeping the Tuist definition simpler and having a plugin (or maybe many plugins, one for each way of grouping them) which adds some syntactic sugar on top of it could be the right choice

**Comment by pepicrft** [comment]
> This makes it clear what the relationship between the two is, allowing for them to be grouped in the same scheme. It also reduces duplication, as omitted fields can be inferred from the parent.

I had a similar thought a while ago but gave up on it because the mental model breaks when a target has more than one parent. I see two non-mutually exclusive solutions:

- An API to declare how targets are glued together under an umbrella scheme. The API can be string-based using pattern matching. We can set one by default and let developers override. For example, the unit tests target of a given target has the same name but is suffixed with `Tests`.
- An abstraction on top of our current primitives to model a particular architecture, [for example microfeatures](https://docs.tuist.io/tutorial/get-started). Plugins are a great place for those.

**Comment by kwridan** [comment]
> I think there are many way of organizing targets/modules and group them together, so it would be hard to make Tuist decide the "correct" way as it might differ among teams.

Agreed, this will be very subjective as well as team or even project dependant _(e.g. a new project, vs migrating an older project)_

> An abstraction on top of our current primitives to model a particular architecture, for example microfeatures. Plugins are a great place for those.

This scales to suit more needs without adding more to Tuist itself. 

The key pieces are (which I believe Tuist offers in some shape or form):

- Ability to control automatic schemes:
  - The ones generated by Tuist  _(1 scheme / target)_ : Can be disabled via `Config.GenerationOption.autogeneratedSchemes()`
  - The ones automatically created by Xcode: `Workspace.GenerationOptions.automaticXcodeSchemes`
- Ability to define custom schemes: `Project.schemes` / `Workspace.schemes`
- Ability to define re-usable scheme templates: `ProjectDescriptionHelpers` / `Plugin`s

Tuist has some room to play with the defaults, but there will always be cases that won't cater for everyone's exact needs :/

**Comment by danyf90** [comment]
Added proposal [here](https://github.com/tuist/tuist/pull/3986)

**Comment by danyf90** [comment]
Implemented and merged, will be released in 3.0
